File: frontend/types/ws.ts
export interface JoinContestArgs {
    contest_id: string
}

export type WebSocketMessage = 
    | {JoinContest: JoinContestArgs}


export type ResponseType = "Log" | "NextQuestion" | "EndContest";

export interface WebSocketResponse {
    data: ResponseData
}

export type ResponsePayload = LogArgs | NextQuestionArgs | EndContestArgs;

export interface LogArgs {
    message: string,
    isError: boolean
}

export interface NextQuestionArgs {
    question_id: string,
    contest_id: string,
    new_rank: number
}

export interface EndContestArgs {
    contest_id: string
}

export interface ResponseData {
    type: ResponseType;
    payload: ResponsePayload
}

File: common/src/types/ws.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::FullContest;



#[derive(Serialize, Deserialize, Debug, Clone)]
pub enum WebSocketMessage {
    StartContest(FullContest),
    JoinContest(JoinContestArgs),

    Debug
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct JoinContestArgs {
    pub contest_id: Uuid
}


#[derive(Debug, Serialize, Deserialize)]
pub struct WebSocketResponse {
    pub data: ResponseData
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(tag = "type", content = "payload")]
pub enum ResponseData {
    Log(LogArgs),
    NextQuestion(NextQuestionArgs),
    EndContest(EndContestArgs)
}

#[derive(Debug, Serialize, Deserialize)]
pub struct LogArgs {
    pub message: String,

    #[serde(rename="isError")]
    pub is_error: bool
}

#[derive(Debug, Serialize, Deserialize)]
pub struct NextQuestionArgs {
    pub question_id: Uuid,
    pub contest_id: Uuid,
    pub new_rank: i32
}

#[derive(Debug, Serialize, Deserialize)]
pub struct EndContestArgs {
    pub contest_id: Uuid
}